apis_core/apis_entities/edit_generic.py

```diff
@@ -34,6 +37,34 @@
     from apis_highlighter.forms import SelectAnnotatorAgreement
 
 
+@functools.lru_cache
+def get_classes_with_allowed_relation_from(entity_name: str):
```

Comment:
Good point — my thinking was that it is really only used in one place, hence putting it next to where it is used. Maybe it belongs in `utils.caching`, as it's similar in nature to other functions there?

Reply:
I think the `utils.caching` module does not make sense itself - it was apparently created in the wrong assumption that neither Python nor Django can do caching themselves. A lot of functions with various usecases do caching, but I don't think they should all be put into a common module. In the long run I'd propose to move the methods from `utils.caching` to separate `utils` modules.

Reply:
Looking at `caching.py` now, it does seem a little weird — the functions are useful, but there doesn't seem to be a good reason to manually cache the results in some global variable. e.g.

```python
def get_all_class_modules_and_names():
    """
    :return: a list of tuples where each contains a str for the app label and a str for the class
    """

    global _contenttype_class_names
    if _contenttype_class_names is None:
        _contenttype_class_names = []
        for cls in get_all_contenttype_classes():
            _contenttype_class_names.append(
                (cls.__module__.split(".")[-2], cls.__name__)
            )

    return _contenttype_class_names
```

Is a better approach just to use `functools.cache`? Like:

```python
@functools.cache
def get_all_class_modules_and_names():
    """
    :return: a list of tuples where each contains a str for the app label and a str for the class
    """

    # global _contenttype_class_names # Dont' need this global
    # if _contenttype_class_names is None: # Or this check
    contenttype_class_names = []
    for cls in get_all_contenttype_classes():
        contenttype_class_names.append(
            (cls.__module__.split(".")[-2], cls.__name__)
        )

    return contenttype_class_names
```

Reply:
> Is a better approach just to use `functools.cache`?

I think so. Or the [django low level cache](https://docs.djangoproject.com/en/4.2/topics/cache/#the-low-level-cache-api) (not sure about the advantages of the one over the other)

Or remove the functionality: the function you quoted seems to be some kind of custom `ContentTypes` manager reimplementation, which, as we now know, uses its own cache- so it might make sense to simply use that (but I haven't looked at the usecases of the function)